// @botpress/cli@1.7.2/dist/code-generation/integration-package/integration-package-definition/states-module.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var states_module_exports = {};
__export(states_module_exports, {
  StateModule: () => StateModule,
  StatesModule: () => StatesModule
});
module.exports = __toCommonJS(states_module_exports);
var import_generators = require("../../generators");
var import_module = require("../../module");
var strings = __toESM(require("../../strings"));
class StateModule extends import_module.Module {
  constructor(name, _state) {
    super({
      path: `${name}.ts`,
      exportName: strings.varName(name)
    });
    this._state = _state;
  }
  async getContent() {
    return (0, import_generators.jsonSchemaToTypescriptZuiSchema)(this._state.schema, this.exportName, {
      type: `"${this._state.type}" as const`
    });
  }
}
class StatesModule extends import_module.ReExportVariableModule {
  constructor(states) {
    super({ exportName: strings.varName("states") });
    for (const [stateName, state] of Object.entries(states)) {
      const module2 = new StateModule(stateName, state);
      this.pushDep(module2);
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  StateModule,
  StatesModule
});
