// @botpress/cli@1.7.2/dist/linter/rulesets/bot.ruleset.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var bot_ruleset_exports = {};
__export(bot_ruleset_exports, {
  BOT_RULESET: () => BOT_RULESET
});
module.exports = __toCommonJS(bot_ruleset_exports);
var import_spectral_functions = require("@stoplight/spectral-functions");
var import_spectral_functions2 = require("../spectral-functions");
const BOT_RULESET = {
  extends: [],
  rules: {
    "event-outputparams-should-have-title": {
      description: "All event output parameters SHOULD have a title",
      message: "{{description}}: {{error}} SHOULD provide a non-empty title by using .title() in its Zod schema",
      severity: "warn",
      given: "$.events[*]..schema.properties[*]",
      then: [
        {
          field: "x-zui.title",
          function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `output parameter "${path.at(-3)}" of event "${path[1]}"`)
        }
      ]
    },
    "event-outputparams-must-have-description": {
      description: "All event output parameters MUST have a description",
      message: "{{description}}: {{error}} SHOULD provide a non-empty description by using .describe() in its Zod schema",
      severity: "error",
      given: "$.events[*]..schema.properties[*]",
      then: [
        {
          field: "description",
          function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `output parameter "${path.at(-2)}" of event "${path[1]}"`)
        }
      ]
    },
    "configuration-fields-must-have-a-title": {
      description: "All configuration fields MUST have a title",
      message: "{{description}}: {{error}} MUST provide a non-empty title by using .title() in its Zod schema",
      severity: "error",
      given: "$.configuration..schema.properties[*].x-zui",
      then: [
        {
          field: "title",
          function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `configuration parameter "${path.at(-3)}"`)
        }
      ]
    },
    "configuration-fields-must-have-a-description": {
      description: "All configuration fields MUST have a description",
      message: "{{description}}: {{error}} MUST provide a non-empty description by using .describe() in its Zod schema",
      severity: "error",
      given: "$.configuration..schema.properties[*]",
      then: [
        {
          field: "description",
          function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `configuration parameter "${path.at(-2)}"`)
        }
      ]
    },
    "user-tags-should-have-a-title": {
      description: "All user tags SHOULD have a title",
      message: "{{description}}: {{error}} SHOULD have a non-empty title",
      severity: "warn",
      given: "$.user.tags[*]",
      then: [
        {
          field: "title",
          function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `user tag "${path[2]}"`)
        }
      ]
    },
    "user-tags-must-have-a-description": {
      description: "All user tags MUST have a description",
      message: "{{description}}: {{error}} MUST have a non-empty description",
      severity: "error",
      given: "$.user.tags[*]",
      then: [
        {
          field: "description",
          function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `user tag "${path[2]}"`)
        }
      ]
    },
    "conversation-tags-should-have-a-title": {
      description: "All conversation tags SHOULD have a title",
      message: "{{description}}: {{error}} SHOULD have a non-empty title",
      severity: "warn",
      given: "$.conversation.tags[*]",
      then: [
        {
          field: "title",
          function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `conversation tag "${path[2]}"`)
        }
      ]
    },
    "conversation-tags-must-have-a-description": {
      description: "All conversation tags MUST have a description",
      message: "{{description}}: {{error}} MUST have a non-empty description",
      severity: "error",
      given: "$.conversation.tags[*]",
      then: [
        {
          field: "description",
          function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `conversation tag "${path[2]}"`)
        }
      ]
    },
    "message-tags-should-have-a-title": {
      description: "All message tags SHOULD have a title",
      message: "{{description}}: {{error}} SHOULD have a non-empty title",
      severity: "warn",
      given: "$.message.tags[*]",
      then: [
        {
          field: "title",
          function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `message tag "${path[2]}"`)
        }
      ]
    },
    "message-tags-must-have-a-description": {
      description: "All message tags MUST have a description",
      message: "{{description}}: {{error}} MUST have a non-empty description",
      severity: "error",
      given: "$.message.tags[*]",
      then: [
        {
          field: "description",
          function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `message tag "${path[2]}"`)
        }
      ]
    },
    "legacy-zui-title-should-be-removed": {
      description: "Legacy ZUI title fields (ui.title) SHOULD be removed. Please use .title() in your Zod schemas instead",
      severity: "error",
      given: "$..ui[*].title",
      then: [{ function: import_spectral_functions.falsy }]
    },
    "legacy-zui-examples-should-be-removed": {
      description: "Legacy ZUI examples fields (ui.examples) SHOULD be removed. There are currently no alternatives",
      severity: "hint",
      given: "$..ui[*].examples",
      then: [{ function: import_spectral_functions.falsy }]
    },
    "state-fields-should-have-title": {
      description: "All state fields SHOULD have a title",
      message: "{{description}}: {{error}} SHOULD provide a non-empty title by using .title() in its Zod schema",
      severity: "warn",
      given: "$.states[*]..schema.properties[*]",
      then: [
        {
          field: "x-zui.title",
          function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `field "${path.at(-3)}" of state "${path[1]}"`)
        }
      ]
    },
    "state-fields-must-have-description": {
      description: "All state fields MUST have a description",
      message: "{{description}}: {{error}} SHOULD provide a non-empty description by using .describe() in its Zod schema",
      severity: "error",
      given: "$.states[*]..schema.properties[*]",
      then: [
        {
          field: "description",
          function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `field "${path.at(-2)}" of state "${path[1]}"`)
        }
      ]
    }
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BOT_RULESET
});
