// @botpress/cli@1.7.2/dist/code-generation/bot-implementation/bot-implementation.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var bot_implementation_exports = {};
__export(bot_implementation_exports, {
  BotImplementationModule: () => BotImplementationModule
});
module.exports = __toCommonJS(bot_implementation_exports);
var consts = __toESM(require("../consts"));
var import_module = require("../module");
var import_bot_typings = require("./bot-typings");
class BotImplementationModule extends import_module.Module {
  _typingsModule;
  constructor(bot) {
    super({
      exportName: "Bot",
      path: consts.INDEX_FILE
    });
    this._typingsModule = new import_bot_typings.BotTypingsModule(bot);
    this._typingsModule.unshift("typings");
    this.pushDep(this._typingsModule);
  }
  async getContent() {
    const typingsImport = this._typingsModule.import(this);
    return [
      consts.GENERATED_HEADER,
      'import * as sdk from "@botpress/sdk"',
      `import * as ${this._typingsModule.name} from "./${typingsImport}"`,
      `export * from "./${typingsImport}"`,
      "",
      `type TBot = sdk.DefaultBot<${this._typingsModule.name}.${this._typingsModule.exportName}>`,
      "",
      "export class Bot extends sdk.Bot<TBot> {}",
      "",
      "// extra types",
      "",
      "type AsyncFunction = (...args: any[]) => Promise<any>",
      "export type EventHandler = Parameters<Bot['event']>[0]",
      "export type EventHandlerProps = Parameters<EventHandler>[0]",
      "export type MessageHandler = Parameters<Bot['message']>[0]",
      "export type MessageHandlerProps = Parameters<MessageHandler>[0]",
      "export type Client = EventHandlerProps['client']",
      "export type ClientOperation = keyof {",
      "  [K in keyof Client as Client[K] extends AsyncFunction ? K : never]: null",
      "}",
      "export type ClientInputs = {",
      "  [K in ClientOperation]: Parameters<Client[K]>[0]",
      "}",
      "export type ClientOutputs = {",
      "  [K in ClientOperation]: Awaited<ReturnType<Client[K]>>",
      "}",
      "// @deprecated",
      "export type BotEvent = EventHandlerProps['event']",
      "// @deprecated",
      "export type BotEvents = {",
      "  [K in BotEvent['type']]: Extract<BotEvent, { type: K }>",
      "}",
      "// @deprecated",
      "export type BotState = ClientOutputs['getState']['state']",
      "// @deprecated",
      "export type BotStates = {",
      "  [K in BotState['name']]: Extract<BotState, { type: K }>['payload']",
      "}"
    ].join("\n");
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BotImplementationModule
});
