// @botpress/cli@1.7.2/dist/command-implementations/add-command.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var add_command_exports = {};
__export(add_command_exports, {
  AddCommand: () => AddCommand
});
module.exports = __toCommonJS(add_command_exports);
var fslib = __toESM(require("fs"));
var pathlib = __toESM(require("path"));
var import_api = require("../api");
var codegen = __toESM(require("../code-generation"));
var consts = __toESM(require("../consts"));
var errors = __toESM(require("../errors"));
var pkgRef = __toESM(require("../package-ref"));
var utils = __toESM(require("../utils"));
var import_global_command = require("./global-command");
var import_project_command = require("./project-command");
class AddCommand extends import_global_command.GlobalCommand {
  async run() {
    const parsedRef = pkgRef.parsePackageRef(this.argv.packageRef);
    if (!parsedRef) {
      throw new errors.InvalidPackageReferenceError(this.argv.packageRef);
    }
    const targetPackage = parsedRef.type === "path" ? await this._findLocalPackage(parsedRef) : await this._findRemotePackage(parsedRef);
    if (!targetPackage) {
      const notFoundMessage = this.argv.packageType ? `Could not find package "${this.argv.packageRef}" of type "${this.argv.packageType}"` : `Could not find package "${this.argv.packageRef}"`;
      throw new errors.BotpressCLIError(notFoundMessage);
    }
    const packageName = targetPackage.name;
    const baseInstallPath = utils.path.absoluteFrom(utils.path.cwd(), this.argv.installPath);
    const packageDirName = utils.casing.to.kebabCase(packageName);
    const installPath = utils.path.join(baseInstallPath, consts.installDirName, packageDirName);
    const alreadyInstalled = fslib.existsSync(installPath);
    if (alreadyInstalled) {
      this.logger.warn(`Package with name "${packageName}" already installed.`);
      const res = await this.prompt.confirm("Do you want to overwrite the existing package?");
      if (!res) {
        this.logger.log("Aborted");
        return;
      }
      await this._uninstall(installPath);
    }
    let files;
    if (targetPackage.type === "integration") {
      files = await codegen.generateIntegrationPackage(targetPackage.pkg);
    } else {
      files = await codegen.generateInterfacePackage(targetPackage.pkg);
    }
    await this._install(installPath, files);
  }
  async _findRemotePackage(ref) {
    const api = await this.ensureLoginAndCreateClient(this.argv);
    if (this._pkgCouldBe("integration")) {
      const integration = await api.findIntegration(ref);
      if (integration) {
        return { type: "integration", name: integration.name, pkg: { source: "remote", integration } };
      }
    }
    if (this._pkgCouldBe("interface")) {
      const intrface = await api.findPublicInterface(ref);
      if (intrface) {
        return { type: "interface", name: intrface.name, pkg: { source: "remote", interface: intrface } };
      }
    }
    return;
  }
  async _findLocalPackage(ref) {
    const absPath = utils.path.absoluteFrom(utils.path.cwd(), ref.path);
    const projectDefinition = await this._readProject(absPath);
    if (this._pkgCouldBe("integration") && projectDefinition?.type === "integration") {
      return {
        type: "integration",
        name: projectDefinition.definition.name,
        pkg: { source: "local", path: absPath }
      };
    }
    if (this._pkgCouldBe("interface") && projectDefinition?.type === "interface") {
      return {
        type: "interface",
        name: projectDefinition.definition.name,
        pkg: { source: "local", path: absPath }
      };
    }
    if (projectDefinition?.type === "bot") {
      throw new errors.BotpressCLIError("Cannot install a bot as a package");
    }
    return;
  }
  async _install(installPath, files) {
    const line = this.logger.line();
    line.started(`Installing ${files.length} files to "${installPath}"`);
    try {
      for (const file of files) {
        const filePath = utils.path.absoluteFrom(installPath, file.path);
        const dirPath = pathlib.dirname(filePath);
        await fslib.promises.mkdir(dirPath, { recursive: true });
        await fslib.promises.writeFile(filePath, file.content);
      }
      line.success(`Installed ${files.length} files to "${installPath}"`);
    } finally {
      line.commit();
    }
  }
  async _uninstall(installPath) {
    await fslib.promises.rm(installPath, { recursive: true });
  }
  async _readProject(workDir) {
    class AnyProjectCommand extends import_project_command.ProjectCommand {
      async run() {
        throw new errors.BotpressCLIError("Not implemented");
      }
      async readProjectDefinitionFromFS() {
        return super.readProjectDefinitionFromFS();
      }
    }
    const cmd = new AnyProjectCommand(import_api.ApiClient, this.prompt, this.logger, {
      ...this.argv,
      workDir
    });
    return cmd.readProjectDefinitionFromFS().catch((thrown) => {
      if (thrown instanceof errors.ProjectDefinitionNotFoundError) {
        return void 0;
      }
      throw thrown;
    });
  }
  _pkgCouldBe = (pkgType) => {
    if (!this.argv.packageType) {
      return true;
    }
    return this.argv.packageType === pkgType;
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AddCommand
});
