// @botpress/cli/dist/linter/rulesets/interface.ruleset.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var interface_ruleset_exports = {};
__export(interface_ruleset_exports, {
  INTERFACE_RULESET: () => INTERFACE_RULESET
});
module.exports = __toCommonJS(interface_ruleset_exports);
var import_spectral_functions = require("@stoplight/spectral-functions");
var import_spectral_functions2 = require("../spectral-functions");
const INTERFACE_RULESET = {
  extends: [],
  rules: {
    "action-inputparams-should-have-a-title": {
      description: "All action input parameters SHOULD have a title",
      message: "{{description}}: {{error}} SHOULD provide a non-empty title by using .title() in its Zod schema",
      severity: "warn",
      given: "$.actions[*].input..schema.properties[*].x-zui",
      then: [
        {
          field: "title",
          function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `input parameter "${path.at(-3)}" of action "${path[1]}"`)
        }
      ]
    },
    "action-inputparams-must-have-a-description": {
      description: "All action input parameters MUST have a description",
      message: "{{description}}: {{error}} MUST provide a non-empty description by using .describe() in its Zod schema",
      severity: "error",
      given: "$.actions[*].input..schema.properties[*]",
      then: [
        {
          field: "description",
          function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `input parameter "${path.at(-2)}" of action "${path[1]}"`)
        }
      ]
    },
    "action-outputparams-should-have-a-title": {
      description: "All action output parameters SHOULD have a title",
      message: "{{description}}: {{error}} SHOULD provide a non-empty title by using .title() in its Zod schema",
      severity: "warn",
      given: "$.actions[*].output..schema.properties[*].x-zui",
      then: [
        {
          field: "title",
          function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `output parameter "${path.at(-3)}" of action "${path[1]}"`)
        }
      ]
    },
    "action-outputparams-must-have-a-description": {
      description: "All action output parameters MUST have a description",
      message: "{{description}}: {{error}} MUST provide a non-empty description by using .describe() in its Zod schema",
      severity: "error",
      given: "$.actions[*].output..schema.properties[*]",
      then: [
        {
          field: "description",
          function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `output parameter "${path.at(-2)}" of action "${path[1]}"`)
        }
      ]
    },
    "event-outputparams-should-have-title": {
      description: "All event output parameters SHOULD have a title",
      message: "{{description}}: {{error}} SHOULD provide a non-empty title by using .title() in its Zod schema",
      severity: "warn",
      given: "$.events[*]..schema.properties[*]",
      then: [
        {
          field: "x-zui.title",
          function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `output parameter "${path.at(-3)}" of event "${path[1]}"`)
        }
      ]
    },
    "event-outputparams-must-have-description": {
      description: "All event output parameters MUST have a description",
      message: "{{description}}: {{error}} SHOULD provide a non-empty description by using .describe() in its Zod schema",
      severity: "error",
      given: "$.events[*]..schema.properties[*]",
      then: [
        {
          field: "description",
          function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `output parameter "${path.at(-2)}" of event "${path[1]}"`)
        }
      ]
    },
    "legacy-zui-title-should-be-removed": {
      description: "Legacy ZUI title fields (ui.title) SHOULD be removed. Please use .title() in your Zod schemas instead",
      severity: "error",
      given: "$..ui[*].title",
      then: [{ function: import_spectral_functions.falsy }]
    },
    "legacy-zui-examples-should-be-removed": {
      description: "Legacy ZUI examples fields (ui.examples) SHOULD be removed. There are currently no alternatives",
      severity: "hint",
      given: "$..ui[*].examples",
      then: [{ function: import_spectral_functions.falsy }]
    },
    "entities-should-have-a-title": {
      description: "All entities SHOULD have a title",
      message: "{{description}}: {{error}} SHOULD have a non-empty title",
      severity: "warn",
      given: "$.entities[*]",
      then: [{ field: "title", function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `entity "${path[1]}"`) }]
    },
    "entities-must-have-a-description": {
      description: "All entities MUST have a description",
      message: "{{description}}: {{error}} MUST have a non-empty description",
      severity: "error",
      given: "$.entities[*]",
      then: [{ field: "description", function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `entity "${path[1]}"`) }]
    },
    "entity-fields-should-have-a-title": {
      description: "All entity fields SHOULD have a title",
      message: "{{description}}: {{error}} SHOULD provide a non-empty title by using .title() in its Zod schema",
      severity: "warn",
      given: "$.entities[*]..schema.properties[*]",
      then: [
        {
          field: "x-zui.title",
          function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `field "${path.at(-3)}" of entity "${path[1]}"`)
        }
      ]
    },
    "entity-fields-must-have-a-description": {
      description: "All entity fields MUST have a description",
      message: "{{description}}: {{error}} MUST provide a non-empty description by using .describe() in its Zod schema",
      severity: "error",
      given: "$.entities[*]..schema.properties[*]",
      then: [
        {
          field: "description",
          function: (0, import_spectral_functions2.truthyWithMessage)(({ path }) => `field "${path.at(-2)}" of entity "${path[1]}"`)
        }
      ]
    }
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  INTERFACE_RULESET
});
