// @botpress/cli/dist/code-generation/plugin-implementation/plugin-typings/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var plugin_typings_exports = {};
__export(plugin_typings_exports, {
  PluginTypingsModule: () => PluginTypingsModule
});
module.exports = __toCommonJS(plugin_typings_exports);
var consts = __toESM(require("../../consts"));
var import_integration_typings = require("../../integration-implementation/integration-typings");
var import_interface_implementation = require("../../interface-implementation");
var import_module = require("../../module");
var import_actions_module = require("./actions-module");
var import_configuration_module = require("./configuration-module");
var import_events_module = require("./events-module");
var import_states_module = require("./states-module");
class PluginIntegrationsModule extends import_module.ReExportTypeModule {
  constructor(plugin) {
    super({
      exportName: "Integrations"
    });
    for (const [alias, integration] of Object.entries(plugin.integrations ?? {})) {
      const integrationModule = new import_integration_typings.IntegrationTypingsModule(integration.definition);
      integrationModule.unshift(alias);
      this.pushDep(integrationModule);
    }
  }
}
class PluginInterfacesModule extends import_module.ReExportTypeModule {
  constructor(plugin) {
    super({
      exportName: "Interfaces"
    });
    for (const [alias, intrface] of Object.entries(plugin.interfaces ?? {})) {
      const interfaceModule = new import_interface_implementation.InterfaceTypingsModule(intrface.definition);
      interfaceModule.unshift(alias);
      this.pushDep(interfaceModule);
    }
  }
}
const _isLocalPluginDefinition = (plugin) => {
  return "props" in plugin;
};
class PluginTypingsModule extends import_module.Module {
  _dependencies;
  constructor(plugin) {
    super({
      exportName: "TPlugin",
      path: consts.INDEX_FILE
    });
    const integrationsModule = _isLocalPluginDefinition(plugin) ? new PluginIntegrationsModule(plugin) : new import_module.SingleFileModule({
      path: consts.INDEX_FILE,
      exportName: "Integrations",
      content: "export type Integrations = {}"
    });
    integrationsModule.unshift("integrations");
    this.pushDep(integrationsModule);
    const interfacesModule = _isLocalPluginDefinition(plugin) ? new PluginInterfacesModule(plugin) : new import_module.SingleFileModule({
      path: consts.INDEX_FILE,
      exportName: "Interfaces",
      content: "export type Interfaces = {}"
    });
    interfacesModule.unshift("interfaces");
    this.pushDep(interfacesModule);
    const defaultConfigModule = new import_configuration_module.DefaultConfigurationModule(plugin.configuration);
    defaultConfigModule.unshift("configuration");
    this.pushDep(defaultConfigModule);
    const eventsModule = new import_events_module.EventsModule(plugin.events ?? {});
    eventsModule.unshift("events");
    this.pushDep(eventsModule);
    const statesModule = new import_states_module.StatesModule(plugin.states ?? {});
    statesModule.unshift("states");
    this.pushDep(statesModule);
    const actionsModule = new import_actions_module.ActionsModule(plugin.actions ?? {});
    actionsModule.unshift("actions");
    this.pushDep(actionsModule);
    this._dependencies = {
      integrationsModule,
      interfacesModule,
      defaultConfigModule,
      eventsModule,
      statesModule,
      actionsModule
    };
  }
  async getContent() {
    const { integrationsModule, interfacesModule, defaultConfigModule, eventsModule, statesModule, actionsModule } = this._dependencies;
    const integrationsImport = integrationsModule.import(this);
    const interfacesImport = interfacesModule.import(this);
    const defaultConfigImport = defaultConfigModule.import(this);
    const eventsImport = eventsModule.import(this);
    const statesImport = statesModule.import(this);
    const actionsImport = actionsModule;
    return [
      consts.GENERATED_HEADER,
      `import * as ${integrationsModule.name} from './${integrationsImport}'`,
      `import * as ${interfacesModule.name} from './${interfacesImport}'`,
      `import * as ${defaultConfigModule.name} from './${defaultConfigImport}'`,
      `import * as ${eventsModule.name} from './${eventsModule.name}'`,
      `import * as ${statesModule.name} from './${statesModule.name}'`,
      `import * as ${actionsModule.name} from './${actionsImport.name}'`,
      "",
      `export * as ${integrationsModule.name} from './${integrationsImport}'`,
      `export * as ${interfacesModule.name} from './${interfacesImport}'`,
      `export * as ${defaultConfigModule.name} from './${defaultConfigImport}'`,
      `export * as ${eventsModule.name} from './${eventsImport}'`,
      `export * as ${statesModule.name} from './${statesImport}'`,
      `export * as ${actionsModule.name} from './${actionsImport.name}'`,
      "",
      "export type TPlugin = {",
      `  integrations: ${integrationsModule.name}.${integrationsModule.exportName}`,
      `  interfaces: ${interfacesModule.name}.${interfacesModule.exportName}`,
      `  configuration: ${defaultConfigModule.name}.${defaultConfigModule.exportName}`,
      `  events: ${eventsModule.name}.${eventsModule.exportName}`,
      `  states: ${statesModule.name}.${statesModule.exportName}`,
      `  actions: ${actionsModule.name}.${actionsModule.exportName}`,
      "}"
    ].join("\n");
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  PluginTypingsModule
});
