// @botpress/cli/templates/empty-plugin/.botpress/implementation/index.ts
/* eslint-disable */
/* tslint:disable */
// This file is generated. Do not edit it manually.

import * as sdk from "@botpress/sdk"
import * as typings from "./typings/index"
export * from "./typings/index"

type TPlugin = sdk.DefaultPlugin<typings.TPlugin>

export class Plugin extends sdk.Plugin<TPlugin> {}

// extra types

type AsyncFunction = (...args: any[]) => Promise<any>
export type EventHandlers = Required<{
  [K in keyof Plugin['eventHandlers']]: NonNullable<Plugin['eventHandlers'][K]>[number]
}>
export type MessageHandlers = Required<{
  [K in keyof Plugin['messageHandlers']]: NonNullable<Plugin['messageHandlers'][K]>[number]
}>

export type MessageHandlerProps = Parameters<MessageHandlers['*']>[0]
export type EventHandlerProps = Parameters<EventHandlers['*']>[0]

export type Client = (MessageHandlerProps | EventHandlerProps)['client']
export type ClientOperation = keyof {
  [K in keyof Client as Client[K] extends AsyncFunction ? K : never]: null
}
export type ClientInputs = {
  [K in ClientOperation]: Parameters<Client[K]>[0]
}
export type ClientOutputs = {
  [K in ClientOperation]: Awaited<ReturnType<Client[K]>>
}
