// @botpress/cli@1.7.2/dist/worker/config.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var config_exports = {};
__export(config_exports, {
  CONFIG_ENV_KEY: () => CONFIG_ENV_KEY,
  configSchema: () => configSchema
});
module.exports = __toCommonJS(config_exports);
var import_sdk = require("@botpress/sdk");
const CONFIG_ENV_KEY = "CHILD_PROCESS_CONFIGURATION";
const configSchema = import_sdk.z.union([
  import_sdk.z.object({
    type: import_sdk.z.literal("code"),
    code: import_sdk.z.string(),
    env: import_sdk.z.record(import_sdk.z.string(), import_sdk.z.string()).optional()
  }),
  import_sdk.z.object({
    type: import_sdk.z.literal("file"),
    file: import_sdk.z.string(),
    env: import_sdk.z.record(import_sdk.z.string(), import_sdk.z.string()).optional()
  })
]);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CONFIG_ENV_KEY,
  configSchema
});
